src/databricks/sql/auth/retry.py

```diff
@@ -0,0 +1,378 @@
+import logging
+import time
+import typing
+from enum import Enum
+from typing import List, Optional, Tuple, Union
+
+from urllib3 import BaseHTTPResponse  # type: ignore
+from urllib3 import Retry
+from urllib3.util.retry import RequestHistory
+
+from databricks.sql.exc import (
+    CursorAlreadyClosedError,
+    MaxRetryDurationError,
+    NonRecoverableNetworkError,
+    SessionAlreadyClosedError,
+    UnsafeToRetryError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class CommandType(Enum):
+    EXECUTE_STATEMENT = "ExecuteStatement"
+    CLOSE_SESSION = "CloseSession"
+    CLOSE_OPERATION = "CloseOperation"
+    OTHER = "Other"
+
+    @classmethod
+    def get(cls, value: str):
+        value_name_map = {i.value: i.name for i in cls}
+        valid_command = value_name_map.get(value, False)
+        if valid_command:
+            return getattr(cls, str(valid_command))
+        else:
+            return cls.OTHER
+
+
+class DatabricksRetryPolicy(Retry):
+    """
+    Implements our v3 retry policy by extending urllib3's robust default retry behaviour.
+
+    See `should_retry()` for details about what we do and do not retry.
+
+    :param delay_min:
+        Float of seconds for the minimum delay between retries. Passed to urllib3 as its
+        backoff_factor.
+
+    :param delay_max:
+        Float of seconds for the maximum delay between retries. Passed to urllib3 as its
+        backoff_max
+
+    :param stop_after_attempts_count:
+        Integer maximum number of attempts that will be retried. Passed to urllib3 as its
+        total.
+
+    :param stop_after_attempts_duration:
+        Float of maximum number of seconds from the beginning of the first request that a
+        request may be retried. This behaviour is not implemented in urllib3.
+
+    :param delay_default:
+        Float of seconds the connector will wait between sucessive GetOperationStatus
+        requests. This parameter is not used to retry failed network requests. We include
+        it in this class to keep all retry behaviour encapsulated in this file.
+
+    :param force_dangerous_codes:
+        List of integer HTTP status codes that the connector will retry, even for dangerous
+        commands like ExecuteStatement. This is passed to urllib3 by extending its status_forcelist
+
+    :param _retry_start_time:
+        Float unix timestamp. Used to monitor the overall request duration across successive
+        retries. Never set this value directly. Use self.start_retry_time() instead. Users
```

Comment:
if these are not set by users, best not to expose them in a public constructor.

Reply:
Fixed in 88e8a111184571f95f21b6e7f39bc5427303e541